File: 프로그래머스/unrated/132265. 롤케이크 자르기/롤케이크 자르기.py
"""
1. Counter 모듈 통해 누적합 구하기
2. 처음엔 한 쪽이 모두 갖고 있다고 가정하고, 누적합 수 관리하며 0이 되면 del 통해 삭제하기
"""
from collections import Counter


def solution(topping):
    answer = 0
    bro = Counter(topping)
    chul = set()
    print(bro)
    
    for i in topping:
        bro[i] -= 1
        if bro[i] == 0:
            del bro[i]
        chul.add(i)
        if len(bro) == len(chul):
            answer += 1
    

    return answer

File: 프로그래머스/2/132265. 롤케이크 자르기/롤케이크 자르기.java
import java.util.*;

class Solution {
    public int solution(int[] topping) {
        HashMap<Integer, Integer> chul = new HashMap<>();
        HashMap<Integer, Integer> brother = new HashMap<>();
        
        for(int ing:topping){
            if(!chul.containsKey(ing)){
                chul.put(ing, 1);
            } else{
                chul.put(ing, chul.get(ing) + 1);
            }
        }
        
        System.out.println(chul);        
        int answer = 0;
        
        for(int ing:topping){
            if(!brother.containsKey(ing)){
                brother.put(ing, 1);                
            } else{
                brother.put(ing, brother.get(ing) + 1);                
            }            
            
            chul.put(ing, chul.get(ing) - 1);                
            if(chul.get(ing) == 0){
                chul.remove(ing);
            }
            
            if(chul.size() == brother.size()){
                answer++;
            }
            // System.out.println("chul : " + chul);
            // System.out.println("brother : " + brother);
        }
        
        
        return answer;
    }
}